sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,231 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text.Json;
+using Azure.Core.Json;
+
+namespace Azure.Core.Dynamic
+{
+    public partial class DynamicData
+    {
+        internal class AllowList
+        {
+            public static void AssertAllowedType<T>(T value)
+            {
+                if (value == null)
+                {
+                    return;
+                }
+
+                if (!IsAllowedValue(value))
+                {
+                    throw new NotSupportedException($"Type is not currently supported: '{value.GetType()}'.");
```

Comment:
I've updated this - I wanted it still to call out the type to clue users in to why their value wasn't allowed -- because somewhere in the object graph there is an unallowed type.  With the update, trying to assign a BinaryData will result in:

```
System.NotSupportedException: 'Assigning this value is not supported, either because its type 'System.BinaryData' is not allowed, or because it contains unallowed types.'
```

And trying to assign a heterogeneous list of Uri subtypes will result in:

```
System.NotSupportedException: 'Assigning this value is not supported, either because its type 'System.Collections.Generic.List`1[System.Uri]' is not allowed, or because it contains unallowed types.'
```

Let me know if you still have concerns about this experience.